ledger/catchupaccessor.go

```diff
@@ -0,0 +1,490 @@
+type CatchpointCatchupAccessor struct {
+	ledger *Ledger
+
+	// log copied from ledger
+	log logging.Logger
+
+	// Prepared SQL statements for fast accounts DB lookups.
+	accountsq *accountsDbQueries
+}
+
+// CatchpointCatchupState is the state of the current catchpoint catchup process
+type CatchpointCatchupState int
+
+const (
+	// CatchpointCatchupStateInactive is the common state for the catchpoint catchup - not active.
+	CatchpointCatchupStateInactive = iota
+	// CatchpointCatchupStateLedgerDownload indicates that we're downloading the ledger
+	CatchpointCatchupStateLedgerDownload
+	// CatchpointCatchupStateLastestBlockDownload indicates that we're download the latest block
+	CatchpointCatchupStateLastestBlockDownload
+	// CatchpointCatchupStateBlocksDownload indicates that we're downloading the blocks prior to the latest one ( total of MaxBalLookback blocks )
+	CatchpointCatchupStateBlocksDownload
+	// CatchpointCatchupStateSwitch indicates that we're switching to use the downloaded ledger/blocks content
+	CatchpointCatchupStateSwitch
+
+	// catchpointCatchupStateLast is the last entry in the CatchpointCatchupState enumeration.
+	catchpointCatchupStateLast = CatchpointCatchupStateSwitch
+)
+
+// MakeCatchpointCatchupAccessor creates a CatchpointCatchupAccessor given a ledger
+func MakeCatchpointCatchupAccessor(ledger *Ledger, log logging.Logger) *CatchpointCatchupAccessor {
+	rdb := ledger.trackerDB().rdb
+	wdb := ledger.trackerDB().wdb
+	accountsq, err := accountsDbInit(rdb.Handle, wdb.Handle)
+	if err != nil {
+		// todo : log error.
+		return nil
+	}
+	return &CatchpointCatchupAccessor{
+		ledger:    ledger,
+		log:       log,
+		accountsq: accountsq,
+	}
+}
+
+// GetState returns the current state of the catchpoint catchup
+func (c *CatchpointCatchupAccessor) GetState(ctx context.Context) (state CatchpointCatchupState, err error) {
+	var istate uint64
+	istate, _, err = c.accountsq.readCatchpointStateUint64(ctx, "catchpointCatchupState")
+	if err != nil {
+		return 0, fmt.Errorf("unable to read catchpoint catchup state 'catchpointCatchupState': %v", err)
+	}
+	state = CatchpointCatchupState(istate)
+	return
+}
+
+// SetState set the state of the catchpoint catchup
+func (c *CatchpointCatchupAccessor) SetState(ctx context.Context, state CatchpointCatchupState) (err error) {
+	if state < CatchpointCatchupStateInactive || state > catchpointCatchupStateLast {
+		return fmt.Errorf("invalid catchpoint catchup state provided : %d", state)
+	}
+	_, err = c.accountsq.writeCatchpointStateUint64(ctx, "catchpointCatchupState", uint64(state))
+	if err != nil {
+		return fmt.Errorf("unable to write catchpoint catchup state 'catchpointCatchupState': %v", err)
+	}
+	return
+}
+
+// GetLabel returns the current catchpoint catchup label
+func (c *CatchpointCatchupAccessor) GetLabel(ctx context.Context) (label string, err error) {
+	label, _, err = c.accountsq.readCatchpointStateString(ctx, "catchpointCatchupLabel")
```

Comment:
probably it is better to make these string literals named constants

Reply:
done.